dap_aria_mapping/pipeline/data_collection/patents_citations.py

```diff
@@ -0,0 +1,268 @@
+
+from metaflow import FlowSpec, step, Parameter
+import pandas as pd
+from google.cloud import bigquery
+
+from dap_aria_mapping.utils.patents import (
+    format_list_of_strings,
+    chunk,
+)
+
+from dap_aria_mapping.utils.conn import est_conn
+from dap_aria_mapping import logger
+
+# 15 bigquery chunks to run in parallel
+CHUNK_SIZE = 15
+
+# early year time point
+EARLY_YEAR = 2007
+
+# late year time point
+LATE_YEAR = 2017
+
+# create connection to bigquery
+conn = est_conn()
+
+
+def patents_backward_citation_query() -> str:
+    """Returns a query that collects patent citation information
+        (i.e. 'patent citations' in Google Patents front end)
+        for a list of focal_ids.
+
+    Returns:
+        str: Base query string.
+    """
+    return "SELECT p.publication_number as focal_id, c.publication_number as patent_citation from `patents-public-data.patents.publications` as p, unnest(citation) c WHERE p.publication_number IN ({})"
+
+
+def patents_forward_citation_query() -> str:
+    """Returns a query that collects patent forward citation information
+    (i.e. 'cited by' in Google Patents front end)
+    for a list of focal_ids.
+
+    Returns:
+        str: Base query string.
+    """
+    return "SELECT c.publication_number as focal_id, p.publication_number as cited_by from `patents-public-data.patents.publications` as p, unnest(citation) c WHERE c.publication_number IN ({})"
+
+
+def chunk_bigquery_query(
+    publication_numbers: List[str],
+    chunk_size: int,
+    base_q: str,
+    production: bool = False,
+) -> List[List[str]]:
+    """Chunk bigquery query into chunk_size number of chunks
+
+    Args:
+        publication_numbers (List[str]): List of publication numbers
+        chunk_size (int): Number of chunks to split the query into
+        base_q (str): Base query string
+        production (bool, optional): If query is in production. Defaults to False.
+
+    Returns:
+        List[List[str]]: List of query strings
+    """
+    ids = publication_numbers if production else publication_numbers[:10]
+
+    publication_numbers_chunks = chunk(ids, chunk_size)
+
+    return [
+        base_q.format(format_list_of_strings(pub_chunk))
+        for pub_chunk in publication_numbers_chunks
+    ]
+
+
+class PatentsCitationsFlow(FlowSpec):
+    production = Parameter("production", help="Run in production?", default=False)
+    chunk_size = Parameter(
+        "chunk_size", help="# of BigQuery queries", default=CHUNK_SIZE
+    )
+
+    @step
+    def start(self):
+        """
+        Starts the flow.
+        """
+        self.next(self.get_focal_ids)
+
+    @step
+    def get_focal_ids(self):
+        """Gets list of focal_ids from patents data"""
+        from dap_aria_mapping.getters.patents import get_patents
+
+        # get list of publication numbers i.e. focal ids from patents data for early and late years
+        logger.info(
+            f"Fetching list of focal ids for year {EARLY_YEAR} and {LATE_YEAR}."
```

Comment:
In the publication citations flow, I only specified a single year which the user has to vary depending on which year they want to collect. On the one hand your way makes more sense from a reproduction view, but I think it's a little confusing to have a year range that only collects for those specific years. I'd go with a single year and run it twice or an alternative could be to provide the param as a list of years.